Tests/kaas/k8s-version-policy/k8s_version_policy.py

```diff
@@ -395,44 +392,32 @@ async def get_k8s_cluster_info(kubeconfig, context=None) -> ClusterInfo:
-    """Check a given K8s cluster version against the list of released versions in order to find out, if the version
+def check_k8s_version_recency(my_version: K8sVersion, releases_data: list[dict], cve_version_list=()) -> bool:
+    """Check a given K8s cluster version against the list of released versions in order to find out if the version
     is an accepted recent version according to the standard."""
-    if cve_version_list is None:
-        cve_version_list = list()
-
-    github_headers = {
-        "Accept": "application/vnd.github+json",
-        "X-GitHub-Api-Version": "2022-11-28"
-    }
-
-    # Request the latest 100 releases (the next are not needed, since these versions are too old)
-    releases_data = requests.get(
-        "https://api.github.com/repos/kubernetes/kubernetes/releases?per_page=100",
-        headers=github_headers,
-    ).json()
 
+    # iterate over all releases in the list, but only look at those whose branch matches
+    # we might break early assuming that the list is sorted somehow, but it is usually
+    # of bounded length (100), and the loop body not very expensive either
     for release_data in releases_data:
         if release_data['draft'] or release_data['prerelease']:
             continue
 
         release = parse_github_release_data(release_data)
-
-        # Check if the minor version is recent, but allow older versions if requested
-        # FIXME: this assumes k8s stays in 1.x version schema :(
-        if release.version.minor >= my_version.minor and not allow_older:
-            if release.age > MINOR_VERSION_CADENCE:
-                return False
-
-        if my_version.branch == release.version.branch and my_version.patch < release.version.patch:
-            if release.age > PATCH_VERSION_CADENCE:
-                return False
-
-            if release.version in cve_version_list and release.age > CVE_VERSION_CADENCE:
-                return False
-
-        if release.version.minor == (my_version.minor + 1) and release.version.patch == 0:
-            break
+        if my_version.branch != release.version.branch:
+            continue
+        if my_version.patch >= release.version.patch:
+            continue
+        # at this point `release` has the same major.minor, but higher patch than `my_version`
+        if release.age > PATCH_VERSION_CADENCE:
+            # whoops, the cluster should have been updated to his (or a higher version) already!
+            return False
+        if release.version in cve_version_list and release.age > CVE_VERSION_CADENCE:
+            # two FIXMEs:
+            # (a) can the `in` ever become true, when we have a version vs a set of ranges
+            # (b) if the release has a CVE, then there is no use if we updated to it?
+            # shouldn't we check for CVEs of my_version and then check whether the new one still has them?
+            return False
```

Comment:
Regarding (b): You are absolutely right, it should be like this:

```suggestion
        if my_version in cve_version_list and release.age > CVE_VERSION_CADENCE:
            # whoops, the cluster version is affected by a critical CVE and
            # should have already been updated before the usual patch cadence
            return False
```

I think we don't need to check if `release.version` is affected, because we don't tell the user to update to a specific version. We just say that the currently used version has a critical CVE. And if the user fails to update to a version that fixes the CVE and runs the script again it will still complain.

Regarding (a): yes, if the version range `cve_version_list` just represents one version (i.e., only lower_version is set), the `in` should work as expected.

Reply:
The standard expressly states (bold by me):

> This time period MUST be even shorter for patches that target critical CVEs (CVSS >= 8).
> It is RECOMMENDED to **provide a new patch version in a 2 day time period after their release**.

A patch version is a version that doesn't contain the CVE, right? If you agree with my interpretation here, I will gladly implement it.

Reply:
(And, while I'm at it, I will also change `their release` to `its release`)

Reply:
Oh, maybe `their release` refers to the release of the CVEs. Still, I assume the patch version should be one that fixes the CVE.

Reply:
> A patch version is a version that doesn't contain the CVE, right? If you agree with my interpretation here, I will gladly implement it.

No, "patch" just refers to the last number Z in the `X.Y.Z` version scheme. I would rather call it the "fixed version" to avoid confusion.

I think you are right: if `my_version` X.Y.Z has a critical CVE, the algorithm currently only checks if an at least 2 days old patch version X.Y.Z+1 is available and fails. But Z+1 doesn't necessarily need to be the fixed version. It could be Z+2 or later which is younger than 2 days.

Furthermore: the standard says *RECOMMENDED*. The script says *ERROR* in this case.

Reply:
You are right about the terminology (patch). Rather, I should have pointed to this quote: `for patches that target critical CVEs` -- but our conclusions are in agreement anyway.